File: app/src/main/java/org/ncssar/rid2caltopo/data/FlightReadiness.kt
package org.ncssar.rid2caltopo.data

import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant

data class FlightReadiness(
    val aircraft: AircraftReadiness,
    val selectedAccessories: Set<String> = emptySet(),
    val payloadDescription: String = "",
    val payloadWeightGrams: Double? = null,
    val pilotJson: String = "{}",
    val serviceJson: String = "{}",
    val rosterFetchedAt: String = "",
    val confirmedAt: String = "",
    val operatingProfileJson: String? = null,
    val configurationVersion: Long = 0
) {
    fun equipmentJSON(): JSONObject = JSONObject().apply {
        put("selectedAccessories", JSONArray(selectedAccessories.toList()))
        put("payloadDescription", payloadDescription)
        put("payloadWeightGrams", payloadWeightGrams ?: JSONObject.NULL)
    }
    fun restoringEquipment(saved: JSONObject): FlightReadiness {
        val ids = saved.optJSONArray("selectedAccessories") ?: JSONArray()
        return copy(selectedAccessories = (0 until ids.length()).map { ids.getString(it) }.toSet(),
            payloadDescription = saved.optString("payloadDescription"),
            payloadWeightGrams = if (saved.isNull("payloadWeightGrams")) null else saved.optDouble("payloadWeightGrams").takeIf { it.isFinite() })
    }
    fun pilotQualificationWarning(): String {
        return "The saved roster does not verify current Part 107 qualifications for this callsign. Review the callsign or update the pilot’s qualifications in Tracker. Recording and publishing remain available."
    }
    fun equipmentSummary(): String {
        val names = aircraft.accessories.filter { it.id in selectedAccessories }.map { it.name }.toMutableList()
        if (payloadDescription.isNotBlank()) names.add(payloadDescription)
        else if (payloadWeightGrams != null) names.add("Payload")
        return names.joinToString(", ").ifEmpty { "No equipment or payload selected" }
    }
    fun toJSON(): JSONObject = JSONObject().apply {
        operatingProfileJson?.let { put("operatingProfile", JSONObject(it)) }
        put("configurationVersion", configurationVersion)
        put("aircraft", aircraft.toJSON()); put("selectedAccessories", JSONArray(selectedAccessories.toList()))
        put("payloadDescription", payloadDescription); put("payloadWeightGrams", payloadWeightGrams ?: JSONObject.NULL)
        put("pilot", JSONObject(pilotJson)); put("service", JSONObject(serviceJson))
        put("rosterFetchedAt", rosterFetchedAt); put("confirmedAt", confirmedAt)
        put("pilotAttribution", if (JSONObject(pilotJson).optString("memberId").isBlank()) "unresolved" else "operator_selected")
        put("takeoffWeightGrams", aircraft.totalWeight(selectedAccessories, payloadDescription, payloadWeightGrams) ?: JSONObject.NULL)
    }
    fun resolvingPilot(callsign: String, roster: JSONArray?): FlightReadiness {
        val matches = (0 until (roster?.length() ?: 0)).map { roster!!.getJSONObject(it) }
            .filter { callsign.isNotBlank() && it.optString("callsign").equals(callsign.trim(), true) }
        return copy(pilotJson = matches.singleOrNull()?.toString() ?: "{}").withReportedPilot(callsign, matches.size == 1)
    }
    fun withAircraft(local: AircraftReadiness?, published: AircraftReadiness?): FlightReadiness {
        val selected = local ?: published ?: aircraft
        return copy(aircraft = selected, selectedAccessories = selectedAccessories.intersect(selected.accessories.map { it.id }.toSet()))
    }
    fun withReportedPilot(callsign: String, matched: Boolean): FlightReadiness {
        val reported = if (matched) JSONObject(pilotJson) else JSONObject()
        reported.put("callsign", callsign.trim())
        return copy(pilotJson = reported.toString())
    }
    fun confirmed(): FlightReadiness = copy(confirmedAt = Instant.now().toString())
}

File: apple/Sources/R2CCore/FlightReadiness.swift
import Foundation

public struct FlightReadiness: Codable, Sendable, Equatable {
    public var aircraft = AircraftReadiness()
    public var selectedAccessories: Set<String> = []
    public var payloadDescription = ""
    public var payloadWeightGrams: Double?
    public var pilotJSON = "{}"
    public var serviceJSON = "{}"
    public var rosterFetchedAt = ""
    public var confirmedAt = ""
    public var operatingProfileJSON: String?
    public var operatingProfileChangesJSON: String?
    public var configurationVersion: Int64 = 0
    public init() {}

    // Remember equipment only; each flight gets fresh pilot, service and authority evidence.
    public var equipmentDictionary: [String: Any] {
        ["selectedAccessories": selectedAccessories.sorted(), "payloadDescription": payloadDescription,
         "payloadWeightGrams": payloadWeightGrams as Any? ?? NSNull()]
    }

    public func restoringEquipment(_ saved: [String: Any]) -> FlightReadiness {
        var copy = self
        copy.selectedAccessories = Set(saved["selectedAccessories"] as? [String] ?? [])
        copy.payloadDescription = saved["payloadDescription"] as? String ?? ""
        copy.payloadWeightGrams = (saved["payloadWeightGrams"] as? NSNumber)?.doubleValue
        return copy
    }

    public var pilotQualificationWarning: String {
        return "The saved roster does not verify current Part 107 qualifications for this callsign. Review the callsign or update the pilot’s qualifications in Tracker. Recording and publishing remain available."
    }

    public var equipmentSummary: String {
        var names = aircraft.accessories.filter { selectedAccessories.contains($0.id) }.map(\.name)
        if !payloadDescription.isEmpty { names.append(payloadDescription) }
        else if payloadWeightGrams != nil { names.append("Payload") }
        return names.isEmpty ? "No equipment or payload selected" : names.joined(separator: ", ")
    }

    public func resolvingPilot(callsign: String, roster: [[String: Any]]) -> FlightReadiness {
        let name = callsign.trimmingCharacters(in: .whitespacesAndNewlines)
        let matches = roster.filter { !name.isEmpty && ($0["callsign"] as? String ?? "").caseInsensitiveCompare(name) == .orderedSame }
        var copy = self
        copy.pilotJSON = matches.count == 1 ? OperatingProfiles.json(matches[0]) : "{}"
        return copy.withReportedPilot(callsign: name, matched: matches.count == 1)
    }

    public func withAircraft(local: AircraftReadiness?, published: AircraftReadiness?) -> FlightReadiness {
        var copy = self
        copy.aircraft = local ?? published ?? aircraft
        copy.selectedAccessories.formIntersection(copy.aircraft.accessories.map(\.id))
        return copy
    }

    public func withReportedPilot(callsign: String, matched: Bool) -> FlightReadiness {
        var copy = self
        var pilot = matched ? ((dictionary["pilot"] as? [String: Any]) ?? [:]) : [:]
        pilot["callsign"] = callsign.trimmingCharacters(in: .whitespacesAndNewlines)
        if let data = try? JSONSerialization.data(withJSONObject: pilot) { copy.pilotJSON = String(decoding: data, as: UTF8.self) }
        return copy
    }

    public var dictionary: [String: Any] {
        let pilot = (try? JSONSerialization.jsonObject(with: Data(pilotJSON.utf8))) as? [String: Any] ?? [:]
        let service = (try? JSONSerialization.jsonObject(with: Data(serviceJSON.utf8))) as? [String: Any] ?? [:]
        var result: [String: Any] = ["configurationVersion": configurationVersion, "aircraft": aircraft.dictionary, "selectedAccessories": selectedAccessories.sorted(),
                "payloadDescription": payloadDescription, "payloadWeightGrams": payloadWeightGrams as Any? ?? NSNull(),
                "pilot": pilot, "service": service, "rosterFetchedAt": rosterFetchedAt, "confirmedAt": confirmedAt,
                "pilotAttribution": pilot["memberId"] == nil ? "unresolved" : "operator_selected",
                "takeoffWeightGrams": aircraft.totalWeight(selected: selectedAccessories, payloadDescription: payloadDescription, payloadGrams: payloadWeightGrams) as Any? ?? NSNull()]
        if let operatingProfileJSON { result["operatingProfile"] = OperatingProfiles.object(operatingProfileJSON) }
        if let operatingProfileChangesJSON { result["operatingProfileChanges"] = (try? JSONSerialization.jsonObject(with: Data(operatingProfileChangesJSON.utf8))) ?? [] }
        return result
    }
}
